Edit compiler manual (#189)
CQCL/pytket: manual/manual_compiler.rst

When applying passes, you may find that you apply some constraint-solving pass to satisfy a particular :py:class:`Predicate`, but then a subsequent pass will invalidate it by, for example, introducing gates of different gate types or changing which qubits interact via multi-qubit gates. To help understand and manage this, each pass has a set of pre-conditions that specify the requirements assumed on the :py:class:`Circuit` in order for the pass to successfully be applied, and a set of post-conditions that specify which :py:class:`Predicate` s are guaranteed to hold for the outputs and which are invalidated or preserved by the pass. These can be viewed in the API reference for each pass.

Users may find it desirable to enforce their own constraints upon circuits they are working with. It is possible to construct a :py:class:`UserDefinedPredicate` in pytket based on a function that returns a True/False value.

> Below is a minimal example where we construct a predicate which checks if our :py:class:`Circuit` contains less than 3 CX gates.

Reviewer: ```suggestion
Below is a minimal example where we construct a predicate which checks if our :py:class:`Circuit` contains fewer than 3 CX gates.
```

Author: Done

.. `FullPeepholeOptimise` kitchen-sink, but assumes a universal quantum computer

> In practice, peephole and structure-preserving optimisations are almost always strictly beneficial to apply, or at least will never increase the size of the :py:class:`Circuit`. The :py:class:`FullPeepholeOptimise` applies Clifford simplifications, commutes single qubit gates to the front of the circuit and applies passes to squash subcircuits up to three qubits. This provides a one-size-approximately-fits-all "kitchen sink" solution to :py:class:`Circuit` optimisation. This assumes no device constraints by default, so will not generally preserve gateset, connectivity, etc.

Reviewer: ```suggestion
In practice, peephole and structure-preserving optimisations are almost always strictly beneficial to apply, or at least will never increase the size of the :py:class:`Circuit`. The :py:class:`FullPeepholeOptimise` pass applies Clifford simplifications, commutes single-qubit gates to the front of the circuit and applies passes to squash subcircuits of up to three qubits. This provides a one-size-approximately-fits-all "kitchen sink" solution to :py:class:`Circuit` optimisation. This assumes no device constraints by default, so will not generally preserve gateset, connectivity, etc.
```

Author: Done

    [tk1(0, 0, 1.5) node[0];, tk1(0, 0, 1.5) node[1];, tk1(0, 0, 1.5) node[2];, tk1(0, 0, 1.5) node[3];, CX node[1], node[0];, tk1(0, 0, 0.5) node[0];, CX node[1], node[2];, CX node[1], node[3];, tk1(0, 0, 0.5) node[2];, tk1(0, 0, 0.5) node[3];, CX node[3], node[4];, CX node[1], node[3];, CX node[3], node[4];, CX node[4], node[3];, CX node[3], node[4];, CX node[3], node[1];]
    9

.. Note::

>     :py:class:`FullPeepholeOptimise` takes an optional allow_swaps argument. This is a Boolean flag to indicate whether :py:class:`FullPeepholeOptimise` should preserve the circuit connectivity or not. If set to False the pass will presrve circuit connectivity but the circuit will general be less optimal than if connectivity was ignored.

Reviewer: ```suggestion
    :py:class:`FullPeepholeOptimise` takes an optional ``allow_swaps`` argument. This is a boolean flag to indicate whether :py:class:`FullPeepholeOptimise` should preserve the circuit connectivity or not. If set to False the pass will preserve circuit connectivity but the circuit will generally be less optimised than if connectivity was ignored.
```

Author: Thanks for spotting the spelling errors... accidentally deactivated the spell checker in my editor 

done.

>     :py:class:`FullPeepholeOptimise` also takes an optional target_2qb_gate argument to specify whether to target the {:py:class:`OpType.TK1`, :py:class:`OpType.CX`} or {:py:class:`OpType.TK1`, :py:class:`OpType.TK2`} gateset.

Reviewer: ```suggestion
    :py:class:`FullPeepholeOptimise` also takes an optional ``target_2qb_gate`` argument to specify whether to target the {:py:class:`OpType.TK1`, :py:class:`OpType.CX`} or {:py:class:`OpType.TK1`, :py:class:`OpType.TK2`} gateset.
```

Author: done

.. Note::

>     Prevous iterations of :py:class:`FullPeepholeOptimise` did not apply the :py:class:`ThreeQubitSquash` pass. There is a :py:class:`PeepholeOptimise2Q` pass which applies the old pass sequence with the :py:class:`ThreeQubitSquash` pass excluded.

Reviewer: ```suggestion
    Prevous versions of :py:class:`FullPeepholeOptimise` did not apply the :py:class:`ThreeQubitSquash` pass. There is a :py:class:`PeepholeOptimise2Q` pass which applies the old pass sequence with the :py:class:`ThreeQubitSquash` pass excluded.
```

Author: done

.. `Backend.default_compilation_pass` gives a recommended compiler pass to solve the backend's constraints with little or light optimisation

> Also in this category of pre-defined sequences, we have the :py:meth:`Backend.default_compilation_pass()` which is run by :py:meth:`Backend.get_compiled_circuit`. These give a recommended compiler pass to solve the :py:class:`Backend` 's constraints with a choice of optimisation levels.

Reviewer: ```suggestion
Also in this category of pre-defined sequences, we have the :py:meth:`Backend.default_compilation_pass()` which is run by :py:meth:`Backend.get_compiled_circuit`. These give a recommended compiler pass to solve the :py:class:`Backend` 's constraints with a choice of optimisation levels.
```
(It might not be obvious what this change is: the space before "'s" was changed to unicode 0x202f NARROW NO-BREAK SPACE.)

Author: I think this should be fixed now. Thanks.

Done.

Reviewer: My bad, I think this may have been correct before -- looks like it was 0x202f and is now gone?

Author: Okay I've changed it back to how it was before.

Reviewer: Hmm, seems to be a normal space again now?

Easy way to fix would be to copy and paste ":py:class:`Predicate` s" from line 23 and edit.

Author: Oh oops you're right. Should be fixed now hopefully.

2                   Extends to more intensive optimisations (those covered by the :py:meth:`FullPeepholeOptimise` pass).
==================  ========================================================================================================

> We will now demonstrate the :py:meth:`default_compilation_pass` with the different levels of optimisation using the IBMQ Quito device. 

Reviewer: ```suggestion
We will now demonstrate the :py:meth:`default_compilation_pass` with the different levels of optimisation using the ``ibmq_quito`` device. 
```

Author: done

    from pytket.extensions.qiskit import IBMQBackend

    circ = Circuit(3) # Define a circuit to be compiled to the backend
    circ.CX(0, 1)
    circ.H(1)
    circ.Rx(0.42, 1)
    circ.S(1)
    circ.CX(0, 2)
    circ.CX(2, 1)
    circ.Z(2)
    circ.Y(1)
    circ.CX(0, 1)
    circ.CX(2, 0)
    circ.measure_all()

    backend = IBMQBackend("ibmq_quito") # Initialise Backend

    print("Total gate count before compilation =", circ.n_gates)
    print("CX count before compilation =",  circ.n_gates_of_type(OpType.CX))

     # Now apply the default_compilation_pass at different levels of optimisation.

    for ol in range(3):
        test_circ = circ.copy()
        backend.default_compilation_pass(optimisation_level=ol).apply(test_circ)
        assert backend.valid_circuit(test_circ)
        print("Optimisation level", ol)
        print("Gates", test_circ.n_gates)
        print("CXs", test_circ.n_gates_of_type(OpType.CX))
    
.. jupyter-output::

    Total gate count before compilation = 13
    CX count before compilation = 5
    Optimisation level 0
    Gates 22
    CXs 8
    Optimisation level 1
    Gates 12
    CXs 5
    Optimisation level 2
    Gates 6
    CXs 1

**Explanation**

> We see that compiling the circuit to IBMQ Quito at optimisation level 0 actaully increases the gate count. This is because IBM Quito has connectivity constraints which require additional CX gates to be added to validate the circuit.

Reviewer: ```suggestion
We see that compiling the circuit to ``ibmq_quito`` at optimisation level 0 actaully increases the gate count. This is because ``ibmq_quito`` has connectivity constraints which require additional CX gates to be added to validate the circuit.
```

Author: done

> The single qubit gates in our circuit also need to be decomposed into the IBM gatset.

Reviewer: ```suggestion
The single-qubit gates in our circuit also need to be decomposed into the IBM gatset.
```

Author: done.

> We see that compiling at optimisation level 1 manages to reduce the CX count to 5. Our connectivity constraints are satisfied without increasing the CX gate count. Single qubit gates are also combined to reduce the overall gate count further.

Reviewer: ```suggestion
We see that compiling at optimisation level 1 manages to reduce the CX count to 5. Our connectivity constraints are satisfied without increasing the CX gate count. Single-qubit gates are also combined to reduce the overall gate count further.
```

Author: done

> Finally we see that the default pass for optimisation level 2 manages to reduce the overall gate count to just 6 with only one CX gate. This is because more intensive optimisations are applied at this level including squashing passes that enable optimal two and three qubit circuits to be synthesised. Applying these more powerful passes comes with a runtime overhead that may be noticebale for larger circuits.

Reviewer: ```suggestion
Finally we see that the default pass for optimisation level 2 manages to reduce the overall gate count to just 6 with only one CX gate. This is because more intensive optimisations are applied at this level including squashing passes that enable optimal two- and three-qubit circuits to be synthesised. Applying these more powerful passes comes with a runtime overhead that may be noticeable for larger circuits.
```

Author: done
